### PR TITLE
Use specific imports instead of importing whole module

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Description

In order to fight circular import more effectively and speed up astroid import time, we need be specific in what we need to import so we can realize a new architecture can be made.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

### CODE REVIEW

File: astroid/bases.py [hunk 1]
```diff
@@ -87,8 +91,13 @@ def _is_property(meth, context=None):
     if not meth.decorators:
         return False
     for decorator in meth.decorators.nodes or ():
-        inferred = helpers.safe_infer(decorator, context=context)
-        if inferred is None or inferred is util.Uninferable:
+        # Circular import
+        from astroid.helpers import (  # pylint: disable=import-outside-toplevel
+            safe_infer,
+        )
```
Comment (cdce8p) [inline astroid/bases.py:97]: We should avoid putting an import inside a `for` loop.

File: astroid/interpreter/objectmodel.py [hunk 2]
```diff
@@ -799,7 +808,9 @@ def attr_fset(self):
 
         func = self._instance
 
-        def find_setter(func: objects.Property) -> Optional[astroid.FunctionDef]:
+        def find_setter(
+            func: Property,  # pylint: disable=used-before-assignment
+        ) -> Optional[astroid.FunctionDef]:
```
Comment (cdce8p) [inline astroid/interpreter/objectmodel.py:813]:
```suggestion
        def find_setter(func: "Property") -> Optional[astroid.FunctionDef]:
```
This current code causes a runtime exception as `Property` isn't defined. Use `"Property"` instead.

**Review by cdce8p — COMMENTED** [review]
Haven't looked at it in detail yet, will do so later. Left a few comments about things I noticed so far.

File: astroid/bases.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply astroid/bases.py:97]: I did that change too fast, you're damn right.

**Comment by Pierre-Sassoulas** [comment]
Yes I didn't know lazy import well when I started that : the import time actually increase with this change which is why I did not ask for a review on this one. Let's keep or extend the lazy import if required. 

I don't like the third option though, because it's very fragile and if you can easily start to have an actual circular dependency and then you suddenly need to do a huge refactor because the problem was not really addressed. I feel that if we stop using module import (``from astroid import a_module`` and use ``from astroid.a_module import WhatWeReallyUse``) its easier to understand where exactly there are circular dependencies. Because you just have to look at the import at the top of the file instead of scanning every imported module use in the whole file to get a mental map of the import "web". 

I agree that trying to touch ``bases -> helpers`` and ``objectmodel -> objects`` is really demoralizing. I tried too, it felt bad 😄 

But I think we have other tools than the three you talked about though. 
- We can burst the ginormous files that are creating circular dependencies 
- Change the architecture when classes that should be together in a module are not
- Create new file for things that should be a dependency of multiple file instead of being in one and being imported from the other files. 
- Use ``TYPE_CHECKING`` guard when the circular import is due to typing. 
- In last resort, we can also change the design

But I feel that we need to be able to understand the structure first and for that having to look at the import instead of having to look at the full file would help a lot.

**Comment by cdce8p** [comment]
> Let's keep or extend the lazy import if required.

Sounds good to me! Do you want to update this PR then?

> I feel that if we stop using module import (`from astroid import a_module` and use `from astroid.a_module import WhatWeReallyUse`) its easier to understand where exactly there are circular dependencies.

👌🏻

> But I think we have other tools than the three you talked about though.

Yeah, there are more options. The three I mentioned are just the once to I found to resolve the immediate issues.

Just a hunch, but I fear that to truly fix the dependency mess, we'll need to unlink the inference functions from the nodes itself. That would be a lot of work. Might be easier to continue using lazy imports instead.

**Comment by Pierre-Sassoulas** [comment]
Yes I'll update this one later :)

**Comment by Pierre-Sassoulas** [comment]
Last MR for 2.7.3, then I'll release it :) I asked for a review but it's the same that you already reviewed only it's rebased and without the lazy import removal.

**Review by cdce8p — APPROVED** [review]
Looks good! Just a tiny comment

File: astroid/bases.py [hunk 3]
```diff
@@ -436,7 +442,7 @@ def _infer_type_new_call(self, caller, context):
         needs to be a tuple of classes
         """
         # pylint: disable=import-outside-toplevel; circular import
-        from astroid.nodes import node_classes
+        from astroid.nodes.node_classes import Pass
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply astroid/bases.py:445]: It's a gamble as long as the circular import are not really fixed, some file would work some won't, (and some will break if we change something else later on) which is why I don't try right now.

**Pull request merged**